Config: Port EXI device and SI device settings to new config system.

AdmiralCurtiss opened the pull request:
Separate PR here because of this snippet of code that I'm not sure how to translate correctly into the new system:

```
    for (unsigned int i = 0; i < SerialInterface::MAX_SI_CHANNELS; ++i)
    {
      int source;
      controls_section->Get(fmt::format("PadType{}", i), &source, -1);
      if (source >= SerialInterface::SIDEVICE_NONE && source < SerialInterface::SIDEVICE_COUNT)
      {
        StartUp.m_SIDevice[i] = static_cast<SerialInterface::SIDevices>(source);
        config_cache.bSetPads[i] = true;
      }
    }
```

This happens while parsing GameINIs and basically means 'if `PadTypeN` is set to a valid value, override `SIDeviceN` with that, otherwise keep the value set by `SIDeviceN`'. I can and have added these to `GetINIToLocationMap()`, but that's not quite the same behavior -- if you set it to an invalid value, Dolphin will now use `SIDEVICE_NONE` rather than the current `SIDeviceN` value. Are we fine with this? If not, how do we adjust the code so that previous behavior still applies?

JosJuice commented on Source/Core/Core/HW/EXI/EXI.cpp:
```diff
@@ -101,7 +103,9 @@ void Init()
-  g_Channels[0]->AddDevice(SConfig::GetInstance().m_EXIDevice[2], 2);  // Serial Port 1
+  g_Channels[0]->AddDevice(
+      static_cast<ExpansionInterface::TEXIDevices>(Config::Get(Config::MAIN_SERIAL_PORT_1)),
+      2);  // Serial Port 1
```
This comment is redundant now (and makes the line wrapping kind of ugly).

JosJuice commented:
I would be fine with this new PadType behavior.

JosJuice commented on Source/Core/Core/HW/SI/SI_DeviceGCController.cpp:
```diff
@@ -33,6 +34,14 @@ CSIDevice_GCController::CSIDevice_GCController(SIDevices device, int device_numb
+  RefreshConfig();
+}
+
+CSIDevice_GCController ::~CSIDevice_GCController()
```
Note sure why lint didn't catch this, but:

```suggestion
CSIDevice_GCController::~CSIDevice_GCController()
```

JosJuice reviewed (commented):
There's a lot of casting going on in this PR. Is there any particular reason why the settings are defined as `int`s instead of the relevant enum type?

AdmiralCurtiss commented:
I was somehow under the impression that you couldn't forward-declare an old C-style enum, but nope, seems to work fine. I've updated the settings to use the enum directly and removed the casts.

leoetlino commented:
For the record, forward declaring an enum is fine as long as you specify the underlying type (which is only possible starting from C++11). Forward declaring an enum class without specifying the underlying type is okay too (int is the default).
